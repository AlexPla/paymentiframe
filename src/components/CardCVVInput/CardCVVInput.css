/*
* DIV
*/
.card-cvv-input__input-group {
    padding: 12px 6px 0;
    box-sizing: border-box;
    margin-top: 25px;
    width: 100%;
}


/*
* INPUT FIELD
*/
.card-cvv-input__input {
    font: inherit;
    font-size: 14px;
    font-weight: 300;
    margin: 0;
    padding: 10px 30px;
    box-sizing: border-box;
    display: block;
    width: 100%;
    border: none;
    color: #000;
    border-bottom: 1px solid #ccc;
    -moz-appearance: textfield;
    line-height: normal;
}

.card-cvv-input__input:focus {
    outline: none;
    border-color: #000;
    transition: all .2s ease;
    -moz-transition: .2s ease all;
    -webkit-transition: all .2s ease;
}

.card-cvv-input__input_invalid {
    border-color: #e55253;
}

.card-cvv-input__input_invalid:focus {
    border-bottom: 1px solid #e55253;
}

/*
* LABEL - PLACEHOLDER
*/
.card-cvv-input__label {
    color: #666;
    font-size: 14px;
    font-weight: 200;
    position: absolute;
    pointer-events: none;
    left: 35px;
    top: 43px;
    transition: all .2s ease;
    -moz-transition: .2s ease all;
    -webkit-transition: all .2s ease;
}

.card-cvv-input__input:focus~.card-cvv-input__label,
.card-cvv-input__input_not-empty~.card-cvv-input__label {
    top: 25px;
    font-size: 10px;
}


/*
* CARD ICON
*/
.card-cvv-input__icon {
    fill: #666;
    position: absolute;
    top: 41px;
    left: 6px;
    cursor: default;
    text-align: center;
}

.card-cvv-input__input:focus~.card-cvv-input__icon {
    fill: #000;
}


/*
* HELP BUTTON
*/
.card-cvv-input__help-button {
    color: #000;
    position: absolute;
    right: 10px;
    top: 45px;
    cursor: pointer;
    font-size: 14px;
    width: 20px;
    height: 20px;
    text-align: center;
    border-radius: 50%;
    background: #cdcdcd;
}


/*
* ERROR MESSAGE
*/
.card-cvv-input__error {
    position: absolute;
    font-size: 10px;
    font-size: .714rem;
    color: #e55253;
    left: 5px;
}