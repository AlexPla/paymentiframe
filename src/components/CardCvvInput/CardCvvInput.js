import React, { Component } from 'react';
import PropTypes from "prop-types";
import copies from '../../copies/cardCvvInput';
import './CardCvvInput.css';

const CVV_LENGTH_DEFAULT = 3;
const CVV_TYPE_DEFAULT = "CVV";

class CardCvvInput extends Component {
  
  constructor(props, context) {
    super(props, context);

    this.state = {
      errorMessage: '',
      errorDisabled: true
    };

    this.lastCardType = CVV_TYPE_DEFAULT;


      this.onInput = this.onInput.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onClearClick = this.onClearClick.bind(this);
  }

  getErrorMessage(cvv) {
    const pattern = new RegExp(/^[0-9]{3,4}$/);

    switch (true) {
      case cvv.length === 0:
        return copies.errors.required[this.props.lang];
      case cvv.length < this.getLengthLimit() || !pattern.test(cvv):
        return "validation failed";
      default:
        return ""
    }
  }

  updateAppState(card_cvv, errorMessage) {
    this.props.updateFields({ card_cvv });
    this.props.updateErrors({
      key: "card_cvv",
      value: !!errorMessage
    });
  }

  getLengthLimit() {
    return this.props.cardType ? this.props.cardType.code.size : CVV_LENGTH_DEFAULT;
  }

  onInput(e) {
    const cvv = e.target.value.trim();
    if (isNaN(cvv) || cvv.length > this.getLengthLimit()) return null;
    const errorMessage = this.getErrorMessage(cvv);
    this.updateAppState(cvv,  errorMessage);
    this.setState({ errorMessage });
  }

  onClearClick(errorDisabled = false) {
    const errorMessage = copies.errors.required[this.props.lang];
    this.updateAppState('', errorMessage);
    this.setState({ errorMessage, errorDisabled });
  }

  onBlur() {
    const cvv = this.props.value;
    const errorMessage = this.getErrorMessage(cvv);
    this.updateAppState(cvv,  errorMessage);
    this.setState({
      errorDisabled: false,
      errorMessage
    });
  }

  render() {
    const { value } = this.props;
    const { errorMessage, errorDisabled } = this.state;
    const isEmpty = (value).toString().length === 0;
    const code = this.props.cardType ? this.props.cardType.code.name : CVV_TYPE_DEFAULT;
    return (
      <div className='card-cvv-input__input-group' >
        <input
          className={ `card-cvv-input__input ${ !isEmpty && 'card-cvv-input__input_not-empty' } ${ !errorDisabled && errorMessage && 'card-cvv-input__input_invalid' }` }
          type='password'
          autoComplete='on'
          name='cardHolderInput'
          value={ value }
          onInput={ this.onInput }
          onBlur={ this.onBlur }
        />
        <label className='card-cvv-input__label'>{ copies[code][this.props.lang] }</label>
        <svg className='card-cvv-input__icon' version='1.1' xmlns='http://www.w3.org/2000/svg' width='20px' height='20px' viewBox='0 0 150 150' preserveAspectRatio='xMidYMid meet'>
          <path d="M129.7125,51.31875 L113.925,51.31875 L113.925,38.690625 C113.925,17.3625 96.46875,0 75.009375,0 C53.540625,0 36.084375,17.3625 36.084375,38.690625 L36.084375,51.31875 L20.296875,51.31875 C9.178125,51.31875 1.9875,58.509375 1.9875,69.6375 L1.9875,131.68125 C1.978125,142.809375 9.178125,150 20.296875,150 L129.7125,150 C140.85,150 148.040625,142.809375 148.040625,131.68125 L148.040625,69.6375 C148.03125,58.509375 140.83125,51.31875 129.7125,51.31875 L129.7125,51.31875 Z M140.1375,131.690625 C140.1375,138.515625 136.528125,142.115625 129.7125,142.115625 L20.296875,142.115625 C13.48125,142.115625 9.88125,138.50625 9.88125,131.690625 L9.88125,69.6375 C9.88125,62.8125 13.48125,59.2125 20.296875,59.2125 L129.7125,59.2125 C136.5375,59.2125 140.1375,62.821875 140.1375,69.6375 L140.1375,131.690625 L140.1375,131.690625 Z M43.978125,38.690625 C43.978125,21.703125 57.890625,7.89375 75.009375,7.89375 C92.1375,7.89375 106.021875,21.703125 106.021875,38.690625 L106.021875,51.31875 L43.978125,51.31875 L43.978125,38.690625 L43.978125,38.690625 Z" />
        </svg>

        { !isEmpty &&
          <svg className='card-cvv-input__clear-button' onClick={ this.onClearClick } version='1.1' xmlns='http://www.w3.org/2000/svg' width='15px' height='15px' viewBox='0 0 150 150' preserveAspectRatio='xMidYMid meet'>
            <path d='M75.253125,1.021875 C34.10625,1.021875 0.76875,34.359375 0.76875,75.50625 C0.76875,116.653125 34.10625,150 75.253125,150 C116.41875,150 149.7375,116.6625 149.7375,75.515625 C149.7375,34.36875 116.41875,1.021875 75.253125,1.021875 L75.253125,1.021875 Z M107.85,97.715625 L97.4625,108.09375 L75.253125,85.884375 L53.053125,108.09375 L42.675,97.715625 L64.875,75.515625 L42.675,53.30625 L53.053125,42.928125 L75.2625,65.1375 L97.4625,42.928125 L107.85,53.30625 L85.640625,75.515625 L107.85,97.715625 L107.85,97.715625 Z' />
          </svg>
        }
        { !errorDisabled && errorMessage &&
          <span className='card-cvv-input__error' > { errorMessage } </span>
        }
      </div>
    );
  }
}

CardCvvInput.propTypes = {
  updateFields: PropTypes.func.isRequired,
  updateErrors: PropTypes.func.isRequired,
  cardType: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.bool
  ]),
  value: PropTypes.string
};

export default CardCvvInput;
