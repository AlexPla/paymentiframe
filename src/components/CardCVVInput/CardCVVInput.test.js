import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from "enzyme-to-json";
import Adapter from 'enzyme-adapter-react-16';

import CardCVVInput from './CardCVVInput';

configure({ adapter: new Adapter() });

let component, input, updateFieldsMock, updateErrorsMock;

describe('Component CardHolderInput:', () => {

  beforeEach(function () {
    updateFieldsMock = jest.fn();
    updateErrorsMock = jest.fn();
    component = mount(<CardCVVInput
      lang={ 'es' }
      updateFields={ updateFieldsMock }
      updateErrors={ updateErrorsMock }
      cardType={ false }
      value={ '' } />);
    input = component.find(".card-cvv-input__input");
  });

  it('should mount', () => {
    expect(component).toBeDefined();
  });

  it('should match snapshot', () => {
    expect(toJson(component, { mode: "deep" })).toMatchSnapshot();
  });

  it('should call updateFields method from props and pass proper object with new field value as an argument', () => {
    input.prop('onInput')({
      target: {
        value: "123"
      }
    });
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ cardCVV: '123' });
  });

  it('should validate input on input and clear state.errorMessage for proper name', () => {
    input.prop('onInput')({
      target: {
        value: "123"
      }
    });
    component.update();
    expect(component.state('errorMessage')).toBe('');
  });

  it('should validate input on input and set state.errorMessage if holder name is empty', () => {
    input.prop('onInput')({
      target: {
        value: ""
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });


  it('should validate input on input and set state.errorMessage if cvv is less then should be', () => {
    input.prop('onInput')({
      target: {
        value: "12"
      }
    });
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should change cvv length according to card type', () => {
    component.setProps({
      cardType: {
        code: {
          name: "CID",
          size: 4
        }
      }
    });
    component.update();
    input.prop('onInput')({
      target: {
        value: "1234"
      }
    });
    component.update();
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ cardCVV: '1234' });
  });

  it('should call updateErrorsMock with false error value if validation pass on user input', () => {
    input.prop('onInput')({
      target: {
        value: "123"
      }
    });
    component.update();
    expect(updateErrorsMock.mock.calls[0][0]).toEqual({ key: 'cardCVV', value: false });
  });

  it('should not display clear button if input value is empty', () => {
    const clearButton = component.find('.card-cvv-input__clear-button');
    expect(clearButton.length).toBe(0);
  });

  it('should display clear button if input value is not empty', () => {
    component.setProps({ value: "123" });
    component.update();
    const clearButton = component.find('.card-cvv-input__clear-button');
    expect(clearButton.length).not.toBe(0);
  });

  it('should clear input on clear button click', () => {
    component.setProps({ value: "123" });
    component.update();
    const clearButton = component.find('.card-cvv-input__clear-button');
    clearButton.prop('onClick')();
    expect(updateFieldsMock.mock.calls[0][0]).toEqual({ cardCVV: '' });
  });

  it('should set state.errorDisabled to false on clear button click', () => {
    component.setProps({ value: "123" });
    component.update();
    const clearButton = component.find('.card-cvv-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(component.state('errorDisabled')).toBe(false);
  });

  it('should set state.errorMessage with error for empty field on clear button click', () => {
    component.setProps({ value: "123" });
    component.update();
    const clearButton = component.find('.card-cvv-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(component.state('errorMessage')).not.toBe('');
  });

  it('should call updateErrorsMock method from props and pass proper object if validation fails on user input or clear button', () => {
    component.setProps({ value: "123" });
    component.update();
    const clearButton = component.find('.card-cvv-input__clear-button');
    clearButton.prop('onClick')();
    component.update();
    expect(updateErrorsMock.mock.calls[0][0]).toEqual({ key: 'cardCVV', value: true });
  });

  it('should change state.errorDisabled to false on blur', () => {
    input.prop('onBlur')();
    component.update();
    expect(component.state('errorDisabled')).toBe(false);
  });
});
