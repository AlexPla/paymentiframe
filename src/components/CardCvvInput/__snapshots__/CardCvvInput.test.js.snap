// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CardHolderInput: should match snapshot 1`] = `
<div
  className="card-cvv-input__input-group"
>
  <input
    autoComplete="on"
    className="card-cvv-input__input false false"
    name="cardHolderInput"
    onBlur={[Function]}
    onInput={[Function]}
    type="password"
    value=""
  />
  <label
    className="card-cvv-input__label"
  >
    CVV
  </label>
  <svg
    className="card-cvv-input__icon"
    height="20px"
    preserveAspectRatio="xMidYMid meet"
    version="1.1"
    viewBox="0 0 150 150"
    width="20px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M129.7125,51.31875 L113.925,51.31875 L113.925,38.690625 C113.925,17.3625 96.46875,0 75.009375,0 C53.540625,0 36.084375,17.3625 36.084375,38.690625 L36.084375,51.31875 L20.296875,51.31875 C9.178125,51.31875 1.9875,58.509375 1.9875,69.6375 L1.9875,131.68125 C1.978125,142.809375 9.178125,150 20.296875,150 L129.7125,150 C140.85,150 148.040625,142.809375 148.040625,131.68125 L148.040625,69.6375 C148.03125,58.509375 140.83125,51.31875 129.7125,51.31875 L129.7125,51.31875 Z M140.1375,131.690625 C140.1375,138.515625 136.528125,142.115625 129.7125,142.115625 L20.296875,142.115625 C13.48125,142.115625 9.88125,138.50625 9.88125,131.690625 L9.88125,69.6375 C9.88125,62.8125 13.48125,59.2125 20.296875,59.2125 L129.7125,59.2125 C136.5375,59.2125 140.1375,62.821875 140.1375,69.6375 L140.1375,131.690625 L140.1375,131.690625 Z M43.978125,38.690625 C43.978125,21.703125 57.890625,7.89375 75.009375,7.89375 C92.1375,7.89375 106.021875,21.703125 106.021875,38.690625 L106.021875,51.31875 L43.978125,51.31875 L43.978125,38.690625 L43.978125,38.690625 Z"
    />
  </svg>
</div>
`;
